lib/src/box.dart

```diff
@@ -1,337 +1,158 @@
+    int missingIdsCount = 0;
+    for (var instPropVals in allPropVals) {
+      if (instPropVals[_modelEntity.idPropName] == null || instPropVals[_modelEntity.idPropName] == 0) {
+        ++missingIdsCount;
+      }
+    }
+
+    // generate new IDs for these instances and set them
+    if (missingIdsCount != 0) {
+      int nextId = 0;
+      Pointer<Uint64> nextIdPtr = Pointer<Uint64>.allocate(count: 1);
+      try {
+        checkObx(bindings.obx_box_ids_for_put(_cBox, missingIdsCount, nextIdPtr));
+        nextId = nextIdPtr.load<int>();
+      } finally {
+        nextIdPtr.free();
+      }
+      for (var instPropVals in allPropVals) {
+        if (instPropVals[_modelEntity.idPropName] == null || instPropVals[_modelEntity.idPropName] == 0) {
+          instPropVals[_modelEntity.idPropName] = nextId++;
         }
-        return ret;
-    }
-
-    get(int id) {
-        Pointer<Pointer<Void>> dataPtr = Pointer<Pointer<Void>>.allocate();
-        Pointer<Int32> sizePtr = Pointer<Int32>.allocate();
-
-        // get element with specified id from database
-        _inReadTransaction(() => checkObx(bindings.obx_box_get(_objectboxBox, id, dataPtr, sizePtr)));        
-        Pointer<Uint8> data = Pointer<Uint8>.fromAddress(dataPtr.load<Pointer<Void>>().address);
-        var size = sizePtr.load<int>();
-
-        // transform bytes from memory to Dart byte list
-        var buffer = _ByteBuffer(data, size);
-        dataPtr.free();
-        sizePtr.free();
-
-        return _unmarshal(buffer);
-    }
+      }
+    }
+
+    // because obx_box_put_many also needs a list of all IDs of the elements to be put into the box,
+    // generate this list now (only needed if not all IDs have been generated)
+    Pointer<Uint64> allIdsMemory = Pointer<Uint64>.allocate(count: objects.length);
+    for (int i = 0; i < allPropVals.length; ++i)
+      allIdsMemory.elementAt(i).store(allPropVals[i][_modelEntity.idPropName] as int);
+
+    // marshal all objects to be put into the box
+    var putObjects = ByteBufferArray(allPropVals.map<ByteBuffer>(_fbManager.marshal).toList()).toOBXBytesArray();
+
+    checkObx(bindings.obx_box_put_many(_cBox, putObjects.ptr, allIdsMemory, _getOBXPutMode(mode)));
+    putObjects.free();
+    allIdsMemory.free();
+    return allPropVals.map((p) => p[_modelEntity.idPropName] as int).toList();
+  }
+
+  get(int id) {
+    Pointer<Pointer<Void>> dataPtr = Pointer<Pointer<Void>>.allocate();
+    Pointer<Int32> sizePtr = Pointer<Int32>.allocate();
+
+    // get element with specified id from database
+    return _store.runInTransaction(TxMode.Read, () {
+      checkObx(bindings.obx_box_get(_cBox, id, dataPtr, sizePtr));
+
+      Pointer<Uint8> data = Pointer<Uint8>.fromAddress(dataPtr.load<Pointer<Void>>().address);
+      var size = sizePtr.load<int>();
+
+      // transform bytes from memory to Dart byte list
+      var buffer = ByteBuffer(data, size);
+      dataPtr.free();
```

Comment:
In various other places the pointers are freed in a finally block. Should this happen here, too?

Reply:
yes, done